### PR TITLE
develops must call analysis

### PR DESCRIPTION
**Pull request by Kevin-TD**
-must call pass currently sloppily pasted onto called methods pass; updates to code structure soon

### CODE REVIEW

File: src/Analysis.cpp
Comment (kelloggm) [inline src/Analysis.cpp:29]: Do the second one: keep them separate, like we discussed in our call today.

File: src/Analysis.cpp
Comment (kelloggm) [inline src/Analysis.cpp:47]: It depends on what you mean by "unknown function" and "realloc". Can you make this a bit more specific so I can be sure I'm answering the right question?

File: src/Analysis.cpp
Comment (kelloggm) [inline src/Analysis.cpp:822]: I think a lot of this function's code is shared with `transfer` in the called-methods analysis. Would it be possible to abstract out a "generic" transfer function that works for both of them, and have it call methods for the details that need to differ? That would be better software engineering than just copying code between the two.

**Review by kelloggm — CHANGES_REQUESTED** [review]
Before I review this further, I want the code to be restructured:
* the two analyses should be split up into different classes that both implement the same interface (which you'll need to do with an abstract class in C++, IIRC)
* code that both analyses need should exist in only one place, and should be called by both. The correct place for such code is the shared abstract class.

File: src/Analysis.cpp
Comment (Kevin-TD) [reply src/Analysis.cpp:47]: Basically, if we first make a call to malloc then call something like realloc, do we add realloc to the must call set or no?

Comment (kelloggm) [reply src/Analysis.cpp:47]: Calling `realloc` doesn't change the need to call `free`, does it? So I'm not sure that it matters (although it does create a new obligation: you need to erase the called methods set at calls to `realloc` or any other method that might re-allocate the pointer)

File: include/PassType.h
Comment (kelloggm) [inline include/PassType.h:31]: Why do these need to be maps from `string->bool`? I would think just keeping the set of names (as strings) would suffice, without the need for the bool. If the bool is really needed, what it represents needs to be explained in a comment.

File: include/PassType.h
Comment (kelloggm) [inline include/PassType.h:41]: Why is running the tests something that a pass can do? I would expect `runOnCFG` (or something similar) as a method within the pass, and then have `runTests` be in some other class that serves exclusively as a test runner.

Put another way, it would be better to put the _testing logic_ in its own class, so that it doesn't get mixed in with the logic for just running the analysis.

File: include/Utils.h
Comment (kelloggm) [inline include/Utils.h:26]: Is this not in the standard library? I would think you could just call some version of `atoi`?

File: run_all.sh
Comment (kelloggm) [inline run_all.sh:11]: Would it be possible to automatically compute this number from the tests themselves, so that you don't need to update it each time?

Or, we could use an alternative design: give each test a unique name (rather than a number), and then run every test in the directory.

File: src/Analysis.cpp
Comment (kelloggm) [inline src/Analysis.cpp:279]: I think there is something missing from this error message (after "on")

File: src/CalledMethods.cpp
Comment (kelloggm) [inline src/CalledMethods.cpp:133]: Is this `if` block really necessary? It's empty.

**Review by kelloggm — CHANGES_REQUESTED** [review]
This is definitely a dramatic improvement. I left a few comments (all I had time for today), but I will take a deeper look on Monday.

Another thing that we should do (not in this PR, but separately) is setup a code formatter to automatically format the code in this repo. I think my preferred choice would be [clang-format](https://clang.llvm.org/docs/ClangFormat.html), but I'm open to suggestions. Ideally, this would be configured so that it either runs automatically on commit, rejects non-formatted code in CI, or something similar (I don't have a strong preference).

File: include/Utils.h
Comment (Kevin-TD) [reply include/Utils.h:26]: I don't think there is; at most there is isdigit but nothing for an entire string

File: include/PassType.h
Comment (Kevin-TD) [reply include/PassType.h:31]: I opted for using maps since I'd imagine once we further build the code, there would be a lot of safe, unsafe, realloc, & memory functions to check for so a string->bool map would help with making lookup times faster

Comment (kelloggm) [reply include/PassType.h:31]: > I opted for using maps since I'd imagine once we further build the code, there would be a lot of safe, unsafe, realloc, & memory functions to check for so a string->bool map would help with making lookup times faster

The potential size of these maps shouldn't have an impact here: I'm asking about the data representation.

In particular, I don't understand the purpose of the `bool`. What does it represent? That is, when is it true and when is it false? It seems to me that a `std::set<std::string>` should be sufficient here: the presence of a function in the set means that it is a function of the named type; not being in the set means that it is not.

File: include/Utils.h
Comment (kelloggm) [reply include/Utils.h:26]: (You'd check its return value to determine if it succeeded, since it is a conversion function.)

**Review by kelloggm — CHANGES_REQUESTED** [review]
This is a more-detailed review of `PassType` and `CalledMethodsPass`. I suspect that many of my comments also apply to `MustCallPass` given its similarity to `CalledMethodsPass`, so I'd like you to go through it after you've read my comments and fix anything that I mentioned in my review of `CalledMethodsPass` before I review it.

File: src/Analysis.cpp
Comment (Kevin-TD) [reply src/Analysis.cpp:279]: Yeah this should just say "identifed and will be ..."

File: src/CalledMethods.cpp
Comment (Kevin-TD) [reply src/CalledMethods.cpp:16]: Yeah this is just a place to hold the name of the pass in the test code (e.g., test1.txt) so the code knows whether a test case is for called methods or must call. I'm thinking of just putting it in a `Constants.h` file instead, like a `const std::string CALLED_METHODS_NAME = "CalledMethods"` and changing this constructor such that the constructor has no arg and it just does `this->passName = CALLED_METHODS_NAME`

File: src/CalledMethods.cpp
Comment (Kevin-TD) [reply src/CalledMethods.cpp:45]: For the called methods pass itself, it does not need to do anything when an allocation function is called. Probably the reason why this comment is still here is that it's a remnant of code that previously checked for an allocation function that's now deleted. This'll be resolved on the next push where we have more generalizations for the transfer function

File: src/CalledMethods.cpp
Comment (Kevin-TD) [reply src/CalledMethods.cpp:19]: Agreed; it was a name that I picked up from the CIS lab code but I think as well it's better to change it to `instruction`

File: src/CalledMethods.cpp
Comment (Kevin-TD) [reply src/CalledMethods.cpp:81]: Agreed; will be doing that

File: src/CalledMethods.cpp
Comment (Kevin-TD) [reply src/CalledMethods.cpp:90]: Pretty much; these are cases where the programmer might call a function that's never defined

File: include/PassType.h
Comment (Kevin-TD) [reply include/PassType.h:23]: So `AliasMap` is a mapping between LLVM IR intermediate variables (e.g., %6, %7, %8) and variables visible in the code (e.g., %str). It's used to find all local must-aliases, though it is not expected to find all of them once we begin testing this code on more realistic, larger code 
A comment in the code explaining this will be added

File: include/PassType.h
Comment (kelloggm) [inline include/PassType.h:11]: "used to final **all** local must-aliases, though it is not expected to find **all** of them" (emphasis mine) is contradictory. I'd remove the first "all"

File: include/PassType.h
Comment (kelloggm) [reply include/PassType.h:22]: How about `DataflowPass`?

File: include/PassType.h
Comment (kelloggm) [inline include/PassType.h:19]: Why are branch names used as keys? This might be because I don't fully understand what LLVM is giving you, but it goes against my expectations that you'd use branch names at all to identify an estimate: I'd expect that there'd be an entry in this map for every program point, instead.

File: include/PassType.h
Comment (kelloggm) [reply include/PassType.h:22]: I'd also accept `GenericPass` or something similar to that if you'd prefer it, but I think `DataflowPass` is more specific.

File: run_all.py
Comment (kelloggm) [inline run_all.py:1]: Use a shebang instead

File: run_all.py
Comment (kelloggm) [inline run_all.py:7]: This is definitely platform- and possibly machine-specific. Use `os.path.split(os.getcwd())[1]` instead.

File: src/PassType.cpp
Comment (kelloggm) [inline src/PassType.cpp:12]: I think this method logically belongs in `TestRunner`, not this class.

File: src/PassType.cpp
Comment (kelloggm) [inline src/PassType.cpp:91]: This method, too, logically seems to be part of `TestRunner` to me.

File: src/PassType.cpp
Comment (kelloggm) [inline src/PassType.cpp:192]: Not needed in this PR, but we should open issues on this repo to resolve both this comment and the one on line 207. For both, the issue should suggest writing a test to check whether these are correct.

**Review by kelloggm — CHANGES_REQUESTED** [review]
These comments are more minor than my last review: this code is much improved. Thanks! I think this should be ready to merge (hopefully) after one more round of changes.

File: run_all.py
Comment (Kevin-TD) [reply run_all.py:1]: What is a shebang?

File: include/PassType.h
Comment (Kevin-TD) [reply include/PassType.h:22]: I like `DataflowPass` so I'll go with that

File: include/PassType.h
Comment (Kevin-TD) [reply include/PassType.h:19]: I think here we can say branch names are program points; each branch name holds the must call/called methods estimate for each variable after all the code within that branch has run.

File: src/PassType.cpp
Comment (Kevin-TD) [reply src/PassType.cpp:192]: Upon testing and seeing what `argName` is at runtime I now realize that it's more apt to check if `argName` is an empty string (empty string meaning it's not in the `aliasedVars` map). `aliasedVars` filters out irrelevant LLVM IR data by virtue of only storing variables names and intermediate variables (e.g., %str, %1, %2). 
Now for the comment on line 207 I'm still unsure, so that'll probably get its own issue, but I don't think we'll need to go back on the comment for line 192 as I'm confident that we just need to check if `argName` is an empty string

File: run_all.py
Comment (kelloggm) [reply run_all.py:1]: https://en.wikipedia.org/wiki/Shebang_(Unix)

Basically, it's a directive to the OS that tells it what program to use to execute a script.

File: test/test15.c
Comment (kelloggm) [inline test/test15.c:15]: These should definitely be in their own header file that a client program (like a test) should include rather than in the tests directly. For now, this is probably fine, but we'll need to do that eventually so if it's easy to do you should do it now.

File: include/Utils.h
Comment (kelloggm) [reply include/Utils.h:26]: This conversation still hasn't been resolved, and should be before we merge.

**Review by kelloggm — APPROVED** [review]
Two very minor comments. If there's nothing controversial in their resolution, I do not need to approve again before this is merged.

File: include/Utils.h
Comment (Kevin-TD) [reply include/Utils.h:26]: So even if we use `strtol`, it wouldn't be a simple, short one-liner, hence why I still think defining a utility `isNumber` function is a good choice here. Since `isNumber` currently look likes: 
`char* endPtr;
  std::strtol(s.c_str(), &endPtr, 10);
  return endPtr != s.c_str() && *endPtr == '\0';`
  I'd rather have it remain a utility function with a descriptive name

Comment (kelloggm) [reply include/Utils.h:26]: Alright, thanks for looking into it. I agree that keeping a descriptive name is a worthwhile goal.

**Pull request merged**
